exasol/driver/websocket.py

```diff
@@ -0,0 +1,556 @@
+            "socket_timeout": 30,
+            "query_timeout": 0,
+            "compression": False,
+            "encryption": tls,
+            "fetch_dict": False,
+            "fetch_mapper": None,
+            "fetch_size_bytes": 5 * 1024 * 1024,
+            "lower_ident": False,
+            "quote_ident": False,
+            "json_lib": "json",
+            "verbose_error": True,
+            "debug": False,
+            "debug_logdir": None,
+            "udf_output_bind_address": None,
+            "udf_output_connect_address": None,
+            "udf_output_dir": None,
+            "http_proxy": None,
+            "client_name": None,
+            "client_version": None,
+            "client_os_username": None,
+            "protocol_version": 3,
+            "websocket_sslopt": None,
+            "access_token": None,
+            "refresh_token": None,
+        }
+        self._connection = None
+
+    def connect(self):
+        """See also :py:meth: `Connection.connect`"""
+        try:
+            self._connection = pyexasol.connect(**self._options)
+        except pyexasol.exceptions.ExaConnectionError as ex:
+            raise Error("Connection failed") from ex
+        except Exception as ex:
+            raise Error() from ex
+        return self
+
+    @property
+    def connection(self):
+        return self._connection
+
+    def close(self):
+        """See also :py:meth: `Connection.close`"""
+        if not self._connection:
+            return
+        try:
+            self._connection.close()
+        except Exception as ex:
+            raise Error() from ex
+
+    @requires_connection
+    def commit(self):
+        """See also :py:meth: `Connection.commit`"""
+        try:
+            self._connection.commit()
+        except Exception as ex:
+            raise Error() from ex
+
+    @requires_connection
+    def rollback(self):
+        """See also :py:meth: `Connection.rollback`"""
+        try:
+            self._connection.rollback()
+        except Exception as ex:
+            raise Error() from ex
+
+    @requires_connection
+    def cursor(self):
+        """See also :py:meth: `Connection.cursor`"""
+        return DefaultCursor(self)
+
+    def __del__(self):
+        self.close()
+
+
+def requires_result(method):
+    """
+    Decorator requires the object to have a result.
+
+    Raises:
+        Error if the cursor object has not produced a result yet.
+    """
+
+    @wraps(method)
+    def wrapper(self, *args, **kwargs):
+        if not self._cursor:
+            raise Error("No result have been produced.")
+        return method(self, *args, **kwargs)
+
+    return wrapper
+
+
+class DefaultCursor:
+    """
+    Implementation of a cursor based on the DefaultConnection.
+
+    For more details see :class: `Cursor` protocol definition.
+    """
+
+    # see https://peps.python.org/pep-0249/#arraysize
+    DBAPI_DEFAULT_ARRAY_SIZE = 1
+
+    def __init__(self, connection):
```

Comment:
Should we annotate the type of `connection` here? 
```suggestion
    def __init__(self, connection: Connection):
```